addons/website/models/website_visitor.py

```diff
@@ -0,0 +1,129 @@
+
+    @api.depends('last_connection_date')
+    def _compute_is_connected(self):
+        for visitor in self:
+            visitor.is_connected = (datetime.now() - visitor.last_connection_date) < timedelta(minutes=5)
+
+    def _encode(self, visitor_id):
+        md5_visitor_id = md5(b"%d%s" % (visitor_id, self._get_key().encode('ascii'))).hexdigest()
+        return "%d-%s" % (visitor_id, md5_visitor_id)
+
+    def _decode(self):
+        """ Opens the cookie, verifies the signature of the visitor
+        :return: visitor_id if the verification passes, None otherwise"""
+        if request:  # To pass the tests (as in tests, users are authenticated BEFORE making requests)
+            cookie_content = request.httprequest.cookies.get('visitor_id')
+            if cookie_content and '-' in cookie_content:
+                visitor_id, md5_visitor_id = cookie_content.split('-', 1)
+                expected_encryped_visitor_id = md5(("%s%s" % (visitor_id, self._get_key())).encode('utf-8')).hexdigest()
+                if md5_visitor_id == expected_encryped_visitor_id:
+                    visitor = int(visitor_id)
+                    return visitor if visitor else None
+        return None
+
+    def _get_key(self):
+        return self.env['ir.config_parameter'].sudo().get_param('database.secret')
+
+    def _handle_visitor_response(self, response):
+        """ Called on dispatch. This will create a website.visitor if the http request object
+        is a tracked website page. Only on tracked page to avoid having too much operations done on every page
+        or other http requests.
+        Note: The side effect is that the last_connection_date is updated ONLY on tracked pages."""
+        website_page = response.qcontext['main_object']
+        if website_page and website_page.is_tracked:
+            visitor = self.env['website.visitor'].browse(self._decode())
+            if not visitor.exists():
+                # If visitor does not exist
+                visitor_sudo = self._create_visitor(website_page.id)
+                sign = self._encode(visitor_sudo.id)
+                response.set_cookie('visitor_id', sign)
+            else:
+                visitor_sudo = visitor.sudo()
+                # Add page even if already in page_ids as checks on relations are done in many2many write method
+                vals = {
+                    'last_connection_date': datetime.now(),
+                    'page_ids': [(0, 0, {'page_id': website_page.id, 'visit_date': datetime.now()})],
+                }
+                # Would need to round on current day midnight and check if today is another day
+                if visitor_sudo.last_connection_date < (datetime.now() - timedelta(seconds=10)):
+                    vals['visit_count'] = visitor_sudo.visit_count + 1
+                visitor_sudo.write(vals)
+        return response
+
+    def _create_visitor(self, website_page_id=False):
```

Comment:
Seems used only once, probably to inline. Moreover website_page_id always exists in the caller.

Reply:
+1 to remove default value for website_page_id